core/src/main/scala/org/apache/spark/deploy/security/HadoopFSDelegationTokenProvider.scala

```diff
@@ -133,3 +133,45 @@ private[deploy] class HadoopFSDelegationTokenProvider
     if (renewIntervals.isEmpty) None else Some(renewIntervals.min)
   }
 }
+
+private[deploy] object HadoopFSDelegationTokenProvider {
+  def hadoopFSsToAccess(
+      sparkConf: SparkConf,
+      hadoopConf: Configuration): Set[FileSystem] = {
+    val filesystemsToAccess = sparkConf.get(KERBEROS_FILESYSTEMS_TO_ACCESS)
+
+    val defaultFS = FileSystem.get(hadoopConf)
+    val master = sparkConf.get("spark.master", null)
+    val stagingFS = if (master != null && master.contains("yarn")) {
+      sparkConf.get(STAGING_DIR)
+        .map(new Path(_).getFileSystem(hadoopConf))
+        .getOrElse(defaultFS)
```

Comment:
Feels weird to have this fall back to `defaultFS` when the default FS is treated separately. An `Option` seems better.

Reply:
What do you mean exactly here? `defaultFS` is `viewfs` in one of the tests. Not checking that would be a behaviour change.

Reply:
You're already adding `defaultFS` explicitly to the final list. So if `STAGING_DIR` is not defined, you'll have `defaultFS` twice. It's ok because the final list is a set, but why would you do that?

Reply:
I see, fixed.